lib/routes/oauth.js

```diff
@@ -70,6 +70,102 @@ module.exports = (log, config, oauthdb) => {
+            redirect_uri: Joi.string()
+              .max(256)
+              .uri({ scheme: ['https'] })
+              .required(),
+            scope: validators.scope.required(),
+            response_type: Joi.string()
+              .valid('code', 'token')
+              .default('code')
+              .required(),
+            state: Joi.string()
+              .max(256)
+              .when('response_type', {
+                is: 'token',
+                then: Joi.optional(),
+                otherwise: Joi.required()
+              }),
+            ttl: Joi.number()
+              .positive()
+              .when('response_type', {
+                is: 'token',
+                then: Joi.optional(),
+                otherwise: Joi.forbidden()
+              }),
+            access_type: Joi.string()
+              .valid('offline', 'online')
+              .default('online')
+              .optional(),
+            code_challenge_method: Joi.string()
+              .valid('S256')
+              .when('response_type', {
+                is: 'code',
+                then: Joi.optional(),
+                otherwise: Joi.forbidden()
+              }),
+            code_challenge: Joi.string()
+              .length(43) // XXX TODO: magic constant === bad
+              .when('response_type', {
+                is: 'code',
+                then: Joi.optional(),
+                otherwise: Joi.forbidden()
+              }),
+            keys_jwe: validators.jwe
+              .when('response_type', {
+                is: 'code',
+                then: Joi.optional(),
+                otherwise: Joi.forbidden()
+              }),
+            acr_values: Joi.string().max(256).optional().allow(null)
+          }
+        },
+        response: {
+          schema: Joi.object().keys({
+            redirect: Joi.string(),
+            code: Joi.string(),
+            state: Joi.string(),
+            access_token: validators.accessToken,
+            refresh_token: validators.refreshToken.optional(),
+            id_token: validators.assertion.optional(),
+            token_type: Joi.string().valid('bearer'),
+            scope: Joi.string().allow(''),
+            auth_at: Joi.number(),
+            expires_in: Joi.number()
+          }).with('access_token', [
+            'token_type',
+            'scope',
+            'auth_at',
+            'expires_in'
+          ]).with('code', [
+            'state',
+            'redirect',
+          ]).without('code', [
+            'access_token',
+            'refresh_token',
+            'id_token'
+          ])
+        }
+      },
+      handler: async function (request) {
+        const sessionToken = request.auth.credentials
+        if (request.payload.response_type === 'code') {
+          return oauthdb.grantAuthorizationCode(sessionToken, request.payload)
+        } else {
+          return oauthdb.grantAuthorizationToken(sessionToken, request.payload)
+        }
```

Comment:
I purposely split this into two separate calls to the backend, even though they go to the same underlying endpoint, just because the combined schema with all its `with` and `without` strains my ability to hold it correctly in my head.  If and when we do this locally rather than shelling out to fxa-oauth-server, they will definitely want to be two separate code paths.